backend/main_test.go

```diff
@@ -476,3 +465,80 @@ func TestImproveBadMethods(t *testing.T) {
 	postresp := makeHTTPRequest(postreq, handler, t)
 	checkResponseStatus(http.StatusMethodNotAllowed, postresp, t)
 }
+
+type failingAuthStub struct{}
+
+func (auth failingAuthStub) authenticate(next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "Auth failed", http.StatusUnauthorized)
+		return
+	}
+}
+
+func TestAuthMiddleware(t *testing.T) {
+	server := Server{store: in_memory_storage.MakeInMemStore()}
+	server.auth = failingAuthStub{}
+	handler := server.makeHandler()
+
+	testHandlerWithAuth := func(httpMethod, target string, body io.Reader) {
+		req := httptest.NewRequest(httpMethod, target, body)
+		resp := makeHTTPRequest(req, handler, t)
+		checkResponseStatus(http.StatusUnauthorized, resp, t)
+	}
+	teamID := "myteam"
+	periodID := "2019q1"
+
+	testHandlerWithAuth(http.MethodGet, "/api/team/"+teamID, nil)
+	testHandlerWithAuth(http.MethodGet, "/api/team/", nil)
+	testHandlerWithAuth(http.MethodPost, "/api/team/", nil)
+	testHandlerWithAuth(http.MethodPut, "/api/team/"+teamID, nil)
+	testHandlerWithAuth(http.MethodGet, "/api/period/"+teamID+"/"+periodID, nil)
+	testHandlerWithAuth(http.MethodGet, "/api/period/"+teamID+"/", nil)
+	testHandlerWithAuth(http.MethodPost, "/api/period/"+teamID+"/", nil)
+	testHandlerWithAuth(http.MethodPut, "/api/period/"+teamID+"/"+periodID, nil)
+
+	getreq := httptest.NewRequest(http.MethodGet, "/improve", nil)
+	getresp := makeHTTPRequest(getreq, handler, t)
+	checkResponseStatus(http.StatusFound, getresp, t)
+}
+
+type AuthClientStub struct{}
+
+func (AuthClientStub) VerifyIDToken(ctx context.Context, idToken string) (*auth.Token, error) {
+	claims := map[string]interface{}{
+		"user_id":        "id123",
+		"email_verified": true,
+		"email":          "test@google.com"}
```

Comment:
Nit: closing `}` on the next line here, with a trailing comma so we don't have to modify the `email` line if we want to add more elements to this struct?

Reply:
I think the trailing comma threw me off a bit, I changed it now.